### DISCUSSION TITLE
LCM sampler support [very fast and low step generation]

### DISCUSSION THREAD

**Discussion by Amin456789** [leejet/stable-diffusion.cpp#76] [Ideas]
HI!
can someone please implement this sampler and make a script to convert its model to ggml please? @ursg @leejet 

u only need 2-4 steps to generate amazing images! we won't be needing for over 20 steps anymore, it works with cpu too, it has its own model so it has to be converted alongside the sampler itself, please see this here:

https://github.com/luosiallen/latent-consistency-model

it will be amazing with cpp they get along a lot :D

kind regards

**Comment by Green-Sky** [comment]
that looks very impressive. 2-4 steps would be huge.

**Reply by Amin456789** [reply]
yes, it will be great if we have it here!

**Comment by leejet** [comment]
I just added support for the Latent Consistency Models Sampler. You can pull the latest code and try the following steps:

1. Use the script at https://github.com/huggingface/diffusers/blob/main/scripts/convert_diffusers_to_original_stable_diffusion.py to convert the safetensors from https://huggingface.co/SimianLuo/LCM_Dreamshaper_v7 in the diffusers format to the original stable diffusion weights.
2. Then, use the script at https://github.com/leejet/stable-diffusion.cpp/blob/master/models/convert.py to convert the original stable diffusion weights into ggml weights.
3. Now, you can use sd.cpp to run Latent Consistency Models. Please note, it's advisable to set `--cfg-scale` to `1.0` instead of the default `7.0`. For `--steps`, a range of `2-8` steps is recommended. Use `--sampling-method` with `lcm` (although `euler_a` should work as well; the difference between them is minimal).

Here's an example:
```
./bin/sd -m ../models/LCM_Dreamshaper_v7_4k-ggml-model-f16.bin -p "Self-portrait oil painting, a beautiful cyborg with golden hair, 8k" --steps 4 --cfg-scale 1.0 --sampling-method lcm
```
![output](https://github.com/leejet/stable-diffusion.cpp/assets/31925346/53f44434-577f-4b9f-b218-a3dcd49db553)

**Reply by dmikey** [reply]
I've never been so excited! This is awesome news!

**Reply by Green-Sky** [reply]
@leejet for some reason, sampling steps take exactly half the time when cfg-scale is exactly 1.0. any other value (0.5, 2.0 ...) results in the same time.
so ~5sec and ~10sec for me.
any idea why?

**Reply by FSSRepo** [reply]
When cfg is 1, the negative prompt is not taken into account, then unet forward isn't executed twice.

**Reply by juniofaathir** [reply]
@leejet do I need to download the whole files on HF LCM dreamshaper for convertion?

**Reply by Green-Sky** [reply]
the 3 subfolders with the 3 safetensors mentioned here are needed
https://github.com/huggingface/diffusers/blob/4e54dfe985293df9e6e86828d7a2763d076879f5/scripts/convert_diffusers_to_original_stable_diffusion.py#L281-L283

**Reply by HoiV** [reply]
Is there a different repo where convert.py exists? I got a 404 error for the above link: https://github.com/leejet/stable-diffusion.cpp/blob/master/models/convert.py. Thank you!

**Reply by Green-Sky** [reply]
the convert python script got replaced with a convert program. (see [examples/convert](https://github.com/leejet/stable-diffusion.cpp/tree/master/examples/convert) )
you can grab precompiled windows binaries in the release section.

**Reply by HoiV** [reply]
Thank you Erik for the prompt reply. I got the conversion going!

**Reply by HoiV** [reply]
Hit the following issue after a successful conversion (this is on Windows 11) from LCM_Dreamshaper_v7_4k.safetensors to LCM_Dreamshaper_v7_4k.ckpt to LCM_Dreamshaper_v7_4k-f16.gguf. Are there additional steps during the conversion that caused the "unknown tensor" issue below?

..\..\build\bin\Release\sd.exe -m LCM_Dreamshaper_v7_4k-f16.gguf -p "Self-portrait oil painting, a beautiful cyborg with golden hair, 8k" --steps 4 --cfg-scale 1.0 --sampling-method lcm
[INFO]  stable-diffusion.cpp:3715 - loading model from 'LCM_Dreamshaper_v7_4k-f16.gguf'
[INFO]  stable-diffusion.cpp:3743 - Stable Diffusion 1.x | LCM_Dreamshaper_v7_4k.ckpt
[INFO]  stable-diffusion.cpp:3751 - model data type: f16
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_out.0.bias' in model file
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_out.0.weight' in model file
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_v.bias' in model file
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_k.bias' in model file
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_k.weight' in model file
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_q.bias' in model file
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_q.weight' in model file
[WARN]  stable-diffusion.cpp:3831 - unknown tensor 'first_stage_model.decoder.mid.attn_1.to_v.weight' in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.k.bias' not in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.k.weight' not in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.proj_out.bias' not in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.proj_out.weight' not in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.q.bias' not in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.q.weight' not in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.v.bias' not in model file
[ERROR] stable-diffusion.cpp:3889 - tensor 'first_stage_model.decoder.mid.attn_1.v.weight' not in model file

**Reply by FSSRepo** [reply]
Where did you get that .ckpt???

> convert the safetensors from https://huggingface.co/SimianLuo/LCM_Dreamshaper_v7 in the diffusers format to the original stable

Should be safetensors models

**Reply by HoiV** [reply]
You are correct - they are safetensors. However using convert.exe straight on the safetensors file generate an empty gguf file (I made sure to sync all modules and rebuild everything and still hit the same issue #92 - converting of safetensors not working):

convert.exe LCM_Dreamshaper_v7_4k.safetensors -t f16
loading model 'LCM_Dreamshaper_v7_4k.safetensors'
model type: checkpoint
Stable Diffusion 1.x - LCM_Dreamshaper_v7_4k.safetensors
preprocessing 0 tensors
using embedded vocab
converting 0 tensors
alphas_cumprod computed

CLIP Model Tensor count: 0
UNET Model Tensor count: 0
VAE Model Tensor count: 0

saving gguf file
model saved 'LCM_Dreamshaper_v7_4k-f16.gguf' correctly.

I then switched to the Python script: [convert_diffusers_to_original_stable_diffusion.py] to get the ckpt file per the instructions on top of the thread, I was able to convert it to gguf format successfully.

**Reply by FSSRepo** [reply]
@HoiV you can try with my PR #88,  I think I already fixed that issue with some changes I made to the converter since I also encountered a safetensor file that prevented me from converting it.

**Reply by HoiV** [reply]
I have picked up this branch: taesd-impl and synced to this commit 2d760b74772b7e86c3d5cf2deb751cfffb1e6c59. 

After rebuilding convert.exe and ran it over the current LCM_Dreamshaper_v7_4k.safetensors file, I no longer get an empty file but the new GGUF file is a lot smaller than the one you shared with us (which worked great by the way - thank you!). 

This is the new GGUF when I used the bits from PR #88:
1,720,980,672 LCM_Dreamshaper_v7_4k-f16.gguf

And I still get a lot of the same errors of tensor ... not in model file.

**Reply by FSSRepo** [reply]
> After rebuilding convert.exe and ran it over the current LCM_Dreamshaper_v7_4k.safetensors file, I no longer get an empty file but the new GGUF file is a lot smaller than the one you shared with us (which worked great by the way - thank you!). 
> 
> This is the new GGUF when I used the bits from PR #88:
> 1,720,980,672 LCM_Dreamshaper_v7_4k-f16.gguf
> 
> And I still get a lot of the same errors of tensor ... not in model file.

This is not a converter error; you're trying to convert a safetensors file that only contains the diffusion model tensors (missing clip and vae tensors), meaning it's incomplete. I also downloaded that mentioned file, thinking it was the complete model. That's the downside of diffuser architecture models—they're delivered in parts instead of a complete file. It's quite confusing.

**Reply by HoiV** [reply]
Thank you for pointing that out. It looks like that safetensors is only the unet portion. How did you get the full set or how did you manage to create that GGUF file from a complete set given there is no release of a safetensors version that includes everything from HF. Thanks again.

**Comment by juniofaathir** [comment]
I keep got this error message:
```FileNotFoundError: [Errno 2] No such file or directory: '/LCM_Dreamshaper_v7/unet/diffusion_pytorch_model.bin'```
Anyone know what happen?

I've downloaded all the 3 folders and the ```.safetensors``` files. All of them I put it in ```/LCM_Dreamshaper_v7``` folder with unet, vae, and text_encoder sub-folders

**Reply by leejet** [reply]
Follow the steps here to try it
https://github.com/leejet/stable-diffusion.cpp/pull/75#issuecomment-1816826249

**Reply by juniofaathir** [reply]
I did. But it keep gives me same error. It always asking for `.bin` files instead of `.safetensors`

**Reply by leejet** [reply]
Which script did you use for the conversion?

**Reply by juniofaathir** [reply]
I used and follow everything from your link above

**Reply by leejet** [reply]
It looks like you didn't specify the correct model path. `/LCM_Dreamshaper_v7` is a directory under the root directory, are you sure it's not `./LCM_Dreamshaper_v7`，a directory under your current working directory?
```python
    # Path for safetensors
    unet_path = osp.join(args.model_path, "unet", "diffusion_pytorch_model.safetensors")
    vae_path = osp.join(args.model_path, "vae", "diffusion_pytorch_model.safetensors")
    text_enc_path = osp.join(args.model_path, "text_encoder", "model.safetensors")

    # Load models from safetensors if it exists, if it doesn't pytorch
    if osp.exists(unet_path):
        unet_state_dict = load_file(unet_path, device="cpu")
    else:
        unet_path = osp.join(args.model_path, "unet", "diffusion_pytorch_model.bin")
        unet_state_dict = torch.load(unet_path, map_location="cpu")
```

**Reply by juniofaathir** [reply]
Sorry, because I'm still confused, as english is not my main language. But, let me give you an idea of where I put the `LCM_Dreamshaper_v7` and all of it's required files:
![Screenshot from 2023-11-19 19-24-23](https://github.com/leejet/stable-diffusion.cpp/assets/114589511/9f881a56-35cc-4c45-9a10-f5ecad704423)


All above is downloaded from SamianLuo's HF repo. And this is my cli, how I run it >> `python3 convert_diffusers_to_original_stable_diffusion.py --model_path /LCM_Dreamshaper_v7 --checkpoint_path /home/mypc --use_safetensors`

**Reply by leejet** [reply]
Replace `/LCM_Dreamshaper_v7` with `./LCM_Dreamshaper_v7`
```sh
python3 convert_diffusers_to_original_stable_diffusion.py --model_path ./LCM_Dreamshaper_v7 --checkpoint_path /home/mypc --use_safetensors
```

**Comment by leejet** [comment]
I just added LoRA support, now we can specify the use of LCM-LoRA through the prompt. Any SD1.x model should be adaptable.
https://github.com/leejet/stable-diffusion.cpp#with-lora

Here's a simple example:
```
./bin/sd -m ../models/v1-5-pruned-emaonly-ggml-model-f16.bin -p "a lovely cat<lora:lcm-lora-sdv1-5:1>" --steps 4 --lora-model-dir ../models -v --cfg-scale 1
```
![image](https://github.com/leejet/stable-diffusion.cpp/assets/31925346/7e724c6f-60c5-438e-a5ea-3ab69ee88205)

**Reply by Green-Sky** [reply]
This works great, thank you very much :)
Did you test how the lora affects the performance? (memory/compute) I wonder if merging the lcm lora into the model beforehand leads to better performance.

**Reply by FSSRepo** [reply]
Only when loaded, the Lora takes a little time, but there shouldn't be any performance losses since it's the same model, just with slightly modified weights.

**Reply by leejet** [reply]
Yes, there will be no performance loss, just a little extra time loading lora from disk and applying lora to model weights.

**Reply by Green-Sky** [reply]
I hate to bother you with this, but will you implement using multiple loras too?

**Reply by Green-Sky** [reply]
Btw, I tried other loras, and they work great too :+1:

**Reply by leejet** [reply]
Support for multiple LoRAs has already been implemented. You can specify multiple LoRAs at the prompt, similar to how it's done in sd-webui.

```
./bin/sd -m ../models/v1-5-pruned-emaonly-ggml-model-f16.bin -p "a lovely cat<lora:lcm-lora-sdv1-5:1><lora:marblesh:0.6>" --steps 4 --lora-model-dir ../models -v --cfg-scale 1
```

**Reply by Green-Sky** [reply]
not quiet, the lora prompt extractor is faulty for me and only extracts the first lora. see https://github.com/leejet/stable-diffusion.cpp/commit/9a9f3daf8e6759fd7b3397007c2195ce18dd431b#r132974149

**Reply by Green-Sky** [reply]
with my fix:
`a lovely cat<lora:lcm-lora-sdv1-5:1><lora:LuisapPixelArtLORA_v1:1.5>`
![output](https://github.com/leejet/stable-diffusion.cpp/assets/2938071/3cf15bad-f60d-40da-816c-32814625eac3)

**Reply by leejet** [reply]
Thanks for the feedback, I have synced your fix to master.

**Comment by Amin456789** [comment]
thanks guys! u rock!
can't wait for SVD and taesd to be released too!

**Comment by HoiV** [comment]
[like]  Hoi Vo reacted to your message:
________________________________

**Comment by Amin456789** [comment]
i think it will be great if someone upload the converted fp16 gguf of lcm model in huggingface and put the link here for everybody

**Reply by FSSRepo** [reply]
I will do it

**Reply by Amin456789** [reply]
thanks a lot!

**Comment by FSSRepo** [comment]
LCM Dreamshaper GGUF (fp16 for now): https://huggingface.co/Steward/lcm-dreamshaper-v7-gguf

**Reply by Amin456789** [reply]
thank u so much!

**Reply by Amin456789** [reply]
@leejet could u please post this model in the read me? this lcm dreamshaper is woking very good and very important, im afraid many people won't see this one

**Reply by leejet** [reply]
This model doesn't have any differences from using safetensors directly.

**Reply by JohnClaw** [reply]
Please, upload to HF f32 version of LCM.gguf. And GGUFs for SD-Turbo and SDXL-Turbo are also needed.

**Comment by JohnClaw** [comment]
> This model doesn't have any differences from using safetensors directly.

LCM.safetensors doesn't work with the latest version of SD.cpp

**Reply by leejet** [reply]
Where did you download 'LCM.safetensors' from? The file 'LCM_Dreamshaper_v7_4k.safetensors' available at 'https://huggingface.co/SimianLuo/LCM_Dreamshaper_v7/blob/main/LCM_Dreamshaper_v7_4k.safetensors' contains only the weights for the UNet model and cannot be used directly.

**Comment by JohnClaw** [comment]
> Where did you download 'LCM.safetensors' from? The file 'LCM_Dreamshaper_v7_4k.safetensors' available at 'https://huggingface.co/SimianLuo/LCM_Dreamshaper_v7/blob/main/LCM_Dreamshaper_v7_4k.safetensors' contains only the weights for the UNet model and cannot be used directly.

Thank you for the answer. Where can i find the suitable LCM.safetensors?

**Reply by dmikey** [reply]
LCM weights can be found here.

https://huggingface.co/latent-consistency/lcm-lora-sdv1-5
https://huggingface.co/latent-consistency/lcm-lora-sdxl/blob/main/pytorch_lora_weights.safetensors

**Comment by JohnClaw** [comment]
> LCM weights can be found here.

These are loras. I meant standalone all-in-one safetensor or ckpt.

**Comment by HoiV** [comment]
Hi John, given we can currently consume .ckpt / .safetensors directly, I used this script (https://github.com/huggingface/diffusers/blob/main/scripts/convert_diffusers_to_original_stable_diffusion.py)   to reconstruct the .ckpt from the original model from https://huggingface.co/SimianLuo/LCM_Dreamshaper_v7.  From there you can go to any quantization you need from sd.cpp.
